### DISCUSSION TITLE
Feature request: var out keyword scoped to If blocks

### DISCUSSION THREAD

**Discussion by michaelolof** [dotnet/csharplang#3468] [Language Ideas]
This proposal is quite simple.
Consider adding a `var out` keyword which creates a variable in the scope of an if condition.

Example
```cs
public static bool IsGreaterThanFive(this int num, out int rtn)
{
  if(num > 5) { 
    rtn = num;
    return true;
  }
  else {
    rtn = 0;
    return false;
  }
}
```

This can be used as follows
```cs
public static Main(string[] args)
{
  if( 20.IsGreaterThanFive(var out value) ) 
  {
    // value exists only in this scope. 
    Console.Writeln( value ) // Prints 20.
  }
}
```

**Safely Unwrapping Optionals**
```cs
public static Main(string[] args)
{
  Option<string> name = Option<string>.Some( "John Doe" );
  
  if( name.Exists(var out value) ) 
  {
    // value here is sure to be "John Doe";
    Console.Writeln( value ) // "John Doe"
  }

  name.Exists( var out value ) // Compiler error. var out modifier is invalid here.
}
```

The `var out` modifier **will only be allowed in an if condition** and must be scoped to the if block.

**Comment by jnm2** [comment]
This seems too arbitrary and too easy to miss.

**Comment by RamType0** [comment]
It could be done by switch.

```C#
switch (20)
            {
                case var value when value > 5:
                    {
                        Console.WriteLine(value);
                        break;
                    }
            }
```

**Comment by jnm2** [comment]
It can also be done with extra braces:

```cs
public static Main(string[] args)
{
    {
        if (20.IsGreaterThanFive(out var value)) 
        {
            Console.WriteLine(value); // Prints 20.
        }
        
        // value exists only in this scope. 
    }
}
```

**Comment by sharwell** [comment]
I'm not opposed to a feature to constrain scope as much as I'm opposed to `out var` and `var out` both being valid with subtly different meanings. This detail is problematic to the point that it will be difficult to have a good discussion about the merits of the underlying request prior to rewriting the proposal to use a different syntax.

**Comment by theunrepentantgeek** [comment]
The design decision that variables declared by **out var** would "leak" out to the enclosing scope was **_extremely_** heavily debated at the time. 

I would anticipate that LDM members would have no desire at all to revisit the issue again, especially given that there were good reasons for the current decision and easy workarounds to limit scope (as detailed above).
